### PR TITLE
Interoperable Interchain Accounts

### LINKED ISSUE #1747: Make ICAs permissionless
https://discord.com/channels/935678348330434570/1070729373830877215/1070729376137748574

ICA address calculation needs to include the origin chain router address and the destination chain ISM address.

ICA Routers should expose opt-in convenience methods that use these from storage (but always allow senders to be explicitly passing those args)

### LINKED ISSUE #1969: Interchain Account optimizations and PI derivations
@yorhodes in https://github.com/hyperlane-xyz/hyperlane-monorepo/pull/1914 :
- pass proxy address to ICA router constructor
- set OwnableMulticall owner (in bytecode) to proxy
- set ICA router implementation (in bytecode) to OwnableMulticall
- assert that proxy address == OwnableMulticall.owner in deployment

@asaj in https://github.com/hyperlane-xyz/hyperlane-monorepo/pull/1832
step 2:
`getRemoteInterchainAccount`:
- derive OwnableMulticall implementation using CREATE1 on remote router address and zero nonce (constructor deployment)
- derive ICA address using MinimalProxy.bytecode(implementation) CREATE2 on remote router address

### PR DESCRIPTION
**Pull request by asaj**
### Description

This PR introduces a version of Interchain Accounts that is fully interoperable with any compatible contract deployed on a remote chain. 

Rather than follow the previous `Router` pattern in which destination contracts were hard coded by the deployer, this version allows the user to optionally specify which destination contract should receive the interchain account message.

In order to make this secure, interchain accounts are additionally name spaced by ISM and origin router.

**Note that this is not feature-complete until paired with a RoutingISM that switches based on the ISM included in the message**

### Drive-by changes

- Creates the `IRouter` and `IHyperlaneConnectionClient` interfaces
- Moves middleware interfaces into their own directory
- Moves middleware libs into their own directory
- Changes `InterchainCallMessage` -> `InterchainQueryMessage` now that it's unused by anything else

### Related issues

- Fixes #1747 
- Fixes #1969 

### Backward compatibility

_Are these changes backward compatible?_

No

_Are there any infrastructure implications, e.g. changes that would prohibit deploying older commits using this infra tooling?_

None

### Testing

_What kind of testing have these changes undergone?_
Unit Tests

### CODE REVIEW

File: solidity/contracts/libs/middleware/InterchainQueryMessage.sol
Comment (chainchampion) [inline solidity/contracts/libs/middleware/InterchainQueryMessage.sol:56]: @asaj why isn't the encoding similar to and as efficient as `InterchainAccountMessage` being used here

Fore example `format` :
```bash
function format(
    bytes32 _sendar,
    CallLib.StaticCallWithCallback[] calldata _calls
){
    return abi.encodePacked(_sender, types, abi.encode(calls));
}
```

Decoding this is as simple as

```bash
        _sender = bytes32(data[0:32]);
        types = MessageTypes(uint8(data[32]));
        calls = abi.decode(data[33:], (StaticCallWithCallback[]));
```

**Review by chainchampion — COMMENTED** [review]
Put a comment regarding encoding in `InterchainQueryMessage`

File: solidity/contracts/middleware/InterchainAccountRouter.sol
Comment (tkporter) [inline solidity/contracts/middleware/InterchainAccountRouter.sol:50]: feels like this event and 3 subsequent ones have maybe more indexed values than needed. Wondering what the motivation is behind it

e.g. here it and RemoteIsmEnrolled feels like `domain` is the only one that should be indexed

And for the overridden ones possibly just `domain` and `owner`

File: solidity/contracts/middleware/InterchainAccountRouter.sol
Comment (tkporter) [reply solidity/contracts/middleware/InterchainAccountRouter.sol:357]: huh this is interesting. I guess they suggest this to avoid some kind of indexing issue?

File: solidity/contracts/middleware/InterchainAccountRouter.sol
Comment (tkporter) [inline solidity/contracts/middleware/InterchainAccountRouter.sol:316]: Also seems to be missing a zero check on the ism

File: typescript/sdk/src/router.ts
Comment (tkporter) [inline typescript/sdk/src/router.ts:45]: is this intentional? not sure I see this in use anywhere

**Review by tkporter — APPROVED** [review]
generally lgtm, some nits / suggestions. Probably best to get a review from yorke or nam too

File: solidity/contracts/libs/Call.sol
Comment (asaj) [reply solidity/contracts/libs/Call.sol:17]: Thought it was cleaner to not have to do Call.build

File: solidity/contracts/libs/middleware/InterchainAccountMessage.sol
Comment (asaj) [reply solidity/contracts/libs/middleware/InterchainAccountMessage.sol:11]: Ah yes ofc

File: solidity/contracts/middleware/InterchainAccountRouter.sol
Comment (asaj) [reply solidity/contracts/middleware/InterchainAccountRouter.sol:159]: Good call

File: solidity/contracts/middleware/InterchainAccountRouter.sol
Comment (asaj) [reply solidity/contracts/middleware/InterchainAccountRouter.sol:318]: I was thinking a 0 value for ISM means defer to the default in the Mailbox, wdyt?

File: solidity/contracts/middleware/InterchainAccountRouter.sol
Comment (asaj) [reply solidity/contracts/middleware/InterchainAccountRouter.sol:414]: Can you share when/why someone would want to do that?

File: solidity/contracts/middleware/InterchainAccountRouter.sol
Comment (asaj) [reply solidity/contracts/middleware/InterchainAccountRouter.sol:270]: Yes ofc, I had copied over from Router.sol

File: solidity/contracts/middleware/InterchainAccountRouter.sol
Comment (asaj) [reply solidity/contracts/middleware/InterchainAccountRouter.sol:423]: Ditto, had left this from before but I should have cleaned it up

File: solidity/contracts/middleware/InterchainAccountRouter.sol
Comment (asaj) [reply solidity/contracts/middleware/InterchainAccountRouter.sol:50]: I agree. Motivation was to keep the event signatures consistent with Router.sol

File: solidity/contracts/middleware/InterchainAccountRouter.sol
Comment (asaj) [reply solidity/contracts/middleware/InterchainAccountRouter.sol:316]: Ditto, the intention was to allow ISM to be zero to defer to the Mailbox default ISM.

Maybe I should add the routing ISM to this PR?

File: typescript/sdk/src/router.ts
Comment (asaj) [reply typescript/sdk/src/router.ts:45]: Hmm yeah must have missed this when I was cleaning up

File: solidity/contracts/middleware/InterchainAccountRouter.sol
Comment (asaj) [reply solidity/contracts/middleware/InterchainAccountRouter.sol:50]: The question here is do we care about consistency with Router.sol

File: solidity/contracts/middleware/InterchainAccountRouter.sol
Comment (tkporter) [reply solidity/contracts/middleware/InterchainAccountRouter.sol:316]: see https://github.com/hyperlane-xyz/hyperlane-monorepo/pull/1832/files#r1134027231

Remind me what the ICA routing ISM looks like?

File: solidity/contracts/middleware/InterchainAccountRouter.sol
Comment (tkporter) [reply solidity/contracts/middleware/InterchainAccountRouter.sol:414]: If you imagine something like Uniswap governance using ICAs, then they wouldn't require a governance vote to deploy an ICA onto a new chain

I could see this being useful for any DAO that may want an ICA on a chain to be deployed (maybe so ownership can be transferred to it, and they don't feel comfortable just relying on `getLocalInterchainAccount`)

I do admit the safest / most conservative action by a DAO would probably be to deploy the ICA themselves so they know for sure that they have ownership over that ICA. But I could see that being really burdensome for voters

File: solidity/contracts/middleware/InterchainAccountRouter.sol
Comment (tkporter) [reply solidity/contracts/middleware/InterchainAccountRouter.sol:50]: is there any tooling built out for Router.sol that we'd want to still be able to use here? If not, my feeling is it doesn't really matter

File: solidity/contracts/libs/Call.sol
Comment (yorhodes) [reply solidity/contracts/libs/Call.sol:17]: please reorder static sized members at beginning
```
to
value
data
```

File: solidity/contracts/middleware/InterchainAccountRouter.sol
Comment (yorhodes) [inline solidity/contracts/middleware/InterchainAccountRouter.sol:221]: maybe @dev that it doesn't need to be onlyRemoteRouter

File: solidity/interfaces/middleware/IInterchainAccountRouter.sol
Comment (yorhodes) [inline solidity/interfaces/middleware/IInterchainAccountRouter.sol:32]: not sure I agree with the name changes here
at some point we should really stabilize the interfaces

File: solidity/contracts/middleware/InterchainAccountRouter.sol
Comment (asaj) [reply solidity/contracts/middleware/InterchainAccountRouter.sol:414]: I'm okay with this

File: solidity/contracts/middleware/InterchainAccountRouter.sol
Comment (asaj) [reply solidity/contracts/middleware/InterchainAccountRouter.sol:50]: Okay I'll change but keep consistent with Router.sol

File: solidity/contracts/middleware/InterchainAccountRouter.sol
Comment (asaj) [reply solidity/contracts/middleware/InterchainAccountRouter.sol:316]: It defers to the ISM specified in the InterchainAccountMessage

File: solidity/contracts/libs/middleware/InterchainAccountMessage.sol
Comment (asaj) [reply solidity/contracts/libs/middleware/InterchainAccountMessage.sol:33]: Confirmed, fixed

File: solidity/contracts/middleware/InterchainAccountRouter.sol
Comment (asaj) [reply solidity/contracts/middleware/InterchainAccountRouter.sol]: I think you may have reviewed an older version?

File: solidity/contracts/middleware/InterchainAccountRouter.sol
Comment (asaj) [reply solidity/contracts/middleware/InterchainAccountRouter.sol]: Yikes, please excuse the brain fart

File: solidity/contracts/libs/OverridableEnumerableMap.sol
Comment (asaj) [reply solidity/contracts/libs/OverridableEnumerableMap.sol:89]: No we definitely don't. What would you propose?

File: solidity/contracts/middleware/InterchainAccountRouter.sol
Comment (asaj) [reply solidity/contracts/middleware/InterchainAccountRouter.sol:225]: Good call

File: solidity/interfaces/middleware/IInterchainAccountRouter.sol
Comment (asaj) [reply solidity/interfaces/middleware/IInterchainAccountRouter.sol:32]: My goal here was to clarify that the function previously known as `getInterchainAccount` could only be used to get the address for contracts on the local chain. Maybe that was clear already, I'm happy to revert.

I do have a slight preference for `callRemote` over `dispatch`, but I don't feel strongly and am happy to discuss

File: solidity/contracts/libs/middleware/InterchainQueryMessage.sol
Comment (asaj) [reply solidity/contracts/libs/middleware/InterchainQueryMessage.sol:56]: I think @yorhodes correctly called out that this is not as gas efficient as just `abi.encode()`

File: solidity/contracts/middleware/InterchainAccountRouter.sol
Comment (asaj) [reply solidity/contracts/middleware/InterchainAccountRouter.sol]: Unfortunately, we don't have the ID until *after* we make the call. Are y'all cool with merging as-is?

File: solidity/contracts/libs/OverridableEnumerableMap.sol
Comment (yorhodes) [reply solidity/contracts/libs/OverridableEnumerableMap.sol:89]: the consumer can always pass as calldata? and they can have a single storage map

File: solidity/interfaces/middleware/IInterchainAccountRouter.sol
Comment (yorhodes) [reply solidity/interfaces/middleware/IInterchainAccountRouter.sol:32]: happy with `callRemote`
can `getInterchainAccount` be used for nonlocal defaults?

Comment (asaj) [reply solidity/interfaces/middleware/IInterchainAccountRouter.sol:32]: Only for EVM chains since it uses CREATE2 for address derivation.

File: solidity/contracts/libs/OverridableEnumerableMap.sol
Comment (asaj) [reply solidity/contracts/libs/OverridableEnumerableMap.sol:89]: Hmm sorry I don't totally follow, let's discuss in person tomorrow?

Comment (asaj) [reply solidity/contracts/libs/OverridableEnumerableMap.sol:89]: Chatted offline and agreed to remove overrides configured in storage; if users want to override they use the other function

File: solidity/contracts/middleware/InterchainAccountRouter.sol
Comment (yorhodes) [inline solidity/contracts/middleware/InterchainAccountRouter.sol:410]: maybe pull this into a CREATE library?

File: solidity/contracts/middleware/InterchainAccountRouter.sol
Comment (yorhodes) [inline solidity/contracts/middleware/InterchainAccountRouter.sol:239]: love it

File: solidity/contracts/middleware/InterchainAccountRouter.sol
Comment (yorhodes) [inline solidity/contracts/middleware/InterchainAccountRouter.sol:213]: add a default for local?

File: solidity/contracts/middleware/InterchainAccountRouter.sol
Comment (yorhodes) [reply solidity/contracts/middleware/InterchainAccountRouter.sol:410]: or at least a comment about what this is

Comment (asaj) [reply solidity/contracts/middleware/InterchainAccountRouter.sol:410]: Definitely deserves a comment

File: solidity/contracts/middleware/InterchainAccountRouter.sol
Comment (asaj) [reply solidity/contracts/middleware/InterchainAccountRouter.sol:213]: Yeah so I thought about this, curious what you think.

This feels fundamentally different from getRemoteInterchainAccount, which uses the local->remote settings to derive an address that would come from a local->remote call.

Here, we would be using local->remote settings to derive an address that would come from a remote->local call. i.e. it's the settings on the remote router that really matter.

The ISM to default to, in particular, feels tricky, as we would want this to match whatever the remote router has. I wasn't imagining enrolling a local default ISM.

Comment (yorhodes) [reply solidity/contracts/middleware/InterchainAccountRouter.sol:223]: happy to use mailbox ism as default or punt

**Review by yorhodes — APPROVED** [review]

**Pull request merged**
